### ISSUE TITLE
Create an abstraction layer for the debugging internals and the UI

### ISSUE THREAD

**Issue by RobertZFord** [dosbox-staging/dosbox-staging#402]
This is likely a niche suggestion, but it may help to create a layer of abstraction between the debugger and the UI.

I have a smaller laptop screen, and it is not large enough to contain the entirety of the debugging UI window.  If there were a layer of abstraction between the Debugger and the UI (say via named pipes or sockets) then it would be possible to write a custom UI for the debugger and interact with it on smaller screens.

(besides, the default debugger interface feels stuffy :) )

If there would be an interest in accepting a PR for this, I could take a stab at it.  I just don't want to waste my time if it's just going to be for myself.

**Comment by kcgen** [comment]
Makes a lot of sense to liberate the debugger's interface from text-mode constraints of dosbox.  In theory, this would open up debugging to rich host-side CLI tools and best-in-class UI frameworks.

Choosing a cross-platform, "future-proof" (as much as possible) GUI framework is another challenge; but would open this area of work up to new pastures of developers and talent. 

Very nice suggestion @RobertZFord !

**Comment by dreamer** [comment]
@RobertZFord What do you think of 477a8688d7c008e9066032a90a38424ee35eb7df? This is a patch imported from https://www.vogons.org/viewtopic.php?t=30538

Overall, if you have some cool ideas for improvements to debugger feature, then go ahead - our backlog is looong, and we appreciate every new pair of eyes/hands/brains helping in pushing the project forward :)

**Comment by RobertZFord** [comment]
Goodness!  That GDB PR looks somewhat heavier than what I had it mind.  I was just looking to make it easier to develop an alternative debugger UI, but if that can be done with GDB, I'll take a look.

My single concern with that would be cross compatibility with Windows clients, but it seems that you can compile GDB for Windows, so that might be doable.  I will take a look at that patch in the morning and see if it works for what I had in mind.

**Comment by kcgen** [comment]
> That GDB PR looks somewhat heavier than what I had it mind.

![Screenshot at 2020-05-24 18-55-42](https://user-images.githubusercontent.com/1557255/82771244-38455080-9df0-11ea-95d5-f70377ae6e5c.png)

Wow.. yes - that's right up there with the 3dfx patch.

**Comment by dreamer** [comment]
Turns out my import of this patch included `.orig` files (remains of SVN conflicts - I think it was my mistake and not accidentally included with the patch). I uploaded fixed version of this patch: b1dc75b642db35bbe47bf82852073af52b4bb2d1 (`vogons/gdbserver-r3778-2`); the patch size went down to +3K,-1K; however it seems like significant part of that 3K is implementation of `poll` function copied from GNU - I guess we don't need it (I am ok with developer options not being cross-platform as long as they work on Linux).

@RobertZFord instead of compiling GDB on Windows, I suggest installing Linux - either natively or via WSL; dealing with mingw environments is generally quite painful.

If you have other ideas about improving the debugging experience for using built-in debugger, then I'm open to them as well :)
